student-records-console-app/studentRecordsConsoleApp.cpp

```diff
@@ -0,0 +1,276 @@
+
+
+// ********************************************** Database Methods **********************************************
+
+// Get all the databases from a given client.
+vector<string> getDatabases(mongocxx::client& client)
+{
+	return client.list_database_names();
+}
+
+
+// ********************************************** Collection Methods **********************************************
+
+// Create a new collection in the given database.
+void createCollection(mongocxx::database& db, const string& collectionName)
+{
+	db.create_collection(collectionName);
+}
+
+// Print the contents of the given collection.
+void printCollection(mongocxx::collection& collection)
+{
+	// Check if collection is empty.
+	if (collection.count_documents({}) == 0)
+	{
+		cout << "Collection is empty." << endl;
+		return;
+	}
+	
+	auto cursor = collection.find({});
+
+	for (auto&& doc : cursor)
+	{
+		cout << bsoncxx::to_json(doc) << endl;
+	}
+}
+
+// ********************************************** Document Methods **********************************************
+
+// Create a document from the given key-value pairs.
+bsoncxx::document::value createDocument(const vector<pair<string, string>>& keyValues)
+{
+	bsoncxx::builder::stream::document document{};
+
+	for (auto& keyValue : keyValues)
+	{
+		document << keyValue.first << keyValue.second;
+	}
+
+	return document << bsoncxx::builder::stream::finalize;
+}
+
+// Insert a document into the given collection.
+void insertDocument(mongocxx::collection& collection, const bsoncxx::document::value& document)
+{
+	collection.insert_one(document.view());
+}
+
+// Delete a document from a given collection.
+void deleteDocument(mongocxx::collection& collection, const bsoncxx::document::value& document)
+{
+	collection.delete_one(document.view());
+}
+
+// Update the document with given key-value pair.
+void updateDocument(mongocxx::collection& collection, const string& key, const string& value, const string& newKey, const string& newValue)
+{
+
+	// Create the query filter
+	auto filter = bsoncxx::builder::stream::document{} << key << value << bsoncxx::builder::stream::finalize;
+
+	//Add query filter argument in find
+	auto cursor = collection.find({ filter });
+
+	for (auto&& doc : cursor)
+	{
+		if (doc[key].get_string().value == value)
+		{
+			collection.update_one(doc, bsoncxx::builder::stream::document{} << "$set" << bsoncxx::builder::stream::open_document << newKey << newValue << bsoncxx::builder::stream::close_document << bsoncxx::builder::stream::finalize);
+		}
+	}
+}
+
+// Find the document with given key-value pair.
+void findDocument(mongocxx::collection& collection, const string& key, const string& value)
+{
+	// Create the query filter
+	auto filter = bsoncxx::builder::stream::document{} << key << value << bsoncxx::builder::stream::finalize;
+	
+	//Add query filter argument in find
+	auto cursor = collection.find({ filter });
+	
+	for (auto&& doc : cursor)
+	{
+		if (doc[key].get_string().value == value)

```

Comment:
Is there a reason why there would be a client-side filter here?

Reply:
Line 106: Find document(s) matching key/value. For example `rollNo: foo`
Line 108: For each of documents returned by the cursor assign to `doc`. Say there are 2 documents matching `rollNo: foo` 
Line 110: If the string value from doc object with key `rollNo` is equal to `foo` print the document 

This condition is redundant because all of the documents returned by the cursor should have `key=rollNo` with the value of `foo`.

Reply:
Removed redundant check.